### PR TITLE
Add Masked Transition component.

### PR DESCRIPTION
**Pull request by jverkoey**
This component makes it possible to present a view controller from a source view, such as a floating action button, using a masked reveal transition.

This component depends on Material Motion's [Transitioning](https://github.com/material-motion/transitioning-objc), [MotionInterchange](https://github.com/material-motion/motion-interchange-objc), and [MotionAnimator](https://github.com/material-motion/motion-animator-objc) libraries.

The component itself is a Transition instance and can be used like this:

    vc.transitionController.transition = MDCMaskedTransition(sourceView: fab)
    present(vc, animated: true)

The component's motion is defined in its motion spec and was extracted directly from the Material motion spec.

The transition supports the following contexts:

- Fullscreen
- Bottom sheet
- Floating card
- Bottom toolbar

### CODE REVIEW

**Comment by jverkoey** [comment]
See #1448 for videos of this component in action.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ajsecord) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: C functions get capitalized (from Google's style guide).

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ajsecord) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: Totally up to you, but I'd `static inline` one-line helper functions.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ajsecord) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: Is the hash mark doing anything here? In general, full sentences with punctuation, please.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ajsecord) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: Nuke blank lines between comments and code, throughout.

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.m
Comment (ajsecord) [inline components/MaskedTransition/src/private/MDCMaskedPresentationController.m]: Capitalize.

Comment (ajsecord) [reply components/MaskedTransition/src/private/MDCMaskedPresentationController.m]: `static UIView *`

File: components/MaskedTransition/src/private/MDCMaskedTransitionMotionSpec.m
Comment (ajsecord) [inline components/MaskedTransition/src/private/MDCMaskedTransitionMotionSpec.m]: We're not in the MDM namespace here, so MDC if any prefix.

Also, macros are nasty—can you name these something self-explanatory, or comment?

**Review by ajsecord — CHANGES_REQUESTED** [review]
README.md with usage information and tests, please.

File: components/MaskedTransition/src/private/MDCMaskedTransitionMotionSpec.m
Comment (jverkoey) [reply components/MaskedTransition/src/private/MDCMaskedTransitionMotionSpec.m]: Renamed.

**Comment by jverkoey** [comment]
Pushed changes.

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.m
Comment (jverkoey) [reply components/MaskedTransition/src/private/MDCMaskedPresentationController.m]: This is an extern so it can't be static.

**Comment by jverkoey** [comment]
Will work on README.md and tests tomorrow.

**Review by ajsecord — COMMENTED** [review]
Changes look good, thanks for making them. Looking forward to the README/tests.

**Comment by jverkoey** [comment]
Added docs.

**Comment by jverkoey** [comment]
In evaluating the potential for tests here I didn't see much that I could test beyond what's already being tested in the dependent libraries. Notably, the Transitioning library includes tests for ensuring that the transition architecture works as expected, and the Motion libraries include tests verifying that animations are created as expected.

I can add a simple UI test that verifies that presenting a view controller with this transition does eventually result in the completion handler being invoked.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: Prefix?

File: components/MaskedTransition/README.md
Comment (ajsecord) [inline components/MaskedTransition/README.md]: `Transitioning` is more or less an implementation detail at this point. Please explain why someone would want to use this from a client's perspective.

**Review by ajsecord — CHANGES_REQUESTED** [review]

File: MaterialComponents.podspec
Comment (ianegordon) [inline MaterialComponents.podspec]: High chance of a name collision.
MDFTransitioning?  MotionTransitioning?

Perhaps not here, but later when you #import <Transitioning/Transitioning>

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: When would we not have a _presentationController?

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.h
Comment (ianegordon) [inline components/MaskedTransition/src/private/MDCMaskedPresentationController.h]: Why do we need to expose the ScrimView?

I would expect this to live privately in MDCMaskedPresentationController.

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.m
Comment (ianegordon) [inline components/MaskedTransition/src/private/MDCMaskedPresentationController.m]: [calculateFrameOfPresentedView copy] ?

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: Consider using MDCHypot(vector.dx, vector.dy) instead of recreating hypot().

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: MDCHypot(initialSourceCenter.x - corner.x, initialSourceCenter.y - corner.y)

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: Won't this overwrite an existing _presentationController?
I wouldn't typically expect to store a reference to a vended PresentationController.

**Review by ianegordon — CHANGES_REQUESTED** [review]
I think a common use case will be someone who wants to take advantage of the custom transition via transitioningDelegate.

// Create filterViewController somewhere
_filterViewController = [[XYZFilterViewController alloc] init...];
_filterViewController.modalPresentationStyle = UIModalPresentationCustom;
_filterViewController.transitioningDelegate = self.maskedTransitionController;

// Handle a tap that summons my view controller
- FABButtonWasTapped {
  [self presentViewController:_filterViewController animated:Yes completion:NULL];
}

Can we expose this functionality?

File: MaterialComponents.podspec
Comment (jverkoey) [reply MaterialComponents.podspec]: Transitioning isn't an MDC library - is its own CocoaPod and is already published to CocoaPods.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: Would rather not add dependencies to other components in the repo.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: It would - the assumption here is that the transition object is only being used with one view controller and can therefor only have one presentation controller.

It may be cleaner if we add a feature to the transition context that exposes the presentation controller if one is available - am happy to add support for that in a follow-up change to Transitioning.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: If we're doing a full-screen presentation (i.e. a non-custom presentation).

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.h
Comment (jverkoey) [reply components/MaskedTransition/src/private/MDCMaskedPresentationController.h]: Ideally the presentation controller does own and manage its scrim, but if the transition doesn't result in the use of a presentation controller (e.g. a non-custom full screen transition) then we have to manage the scrim creation and management in the transition.

File: MaterialComponents.podspec
Comment (ianegordon) [reply MaterialComponents.podspec]: I knew that,... because you already told me.  Apologies, will add caffeine.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: We actually created MDCMath to avoid this very issue.  (A developer re-implementing common math functions)

That said, I can see how LengthOfVector() would be insightful when reading code.
However, still recommend hypot() over rolling your own.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: I'm not sure I follow, your transitioning delegate will only get called if your viewcontroller.modalPresentationStyle is UIModalPresentationCustom.
To put it another way, how would we get our transitioning delegate (and eventually presentation controller) without setting the presentation style to custom?

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: https://github.com/material-motion/transitioning-objc/issues/25

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: The transitioning delegate is queried for an animation and interaction controller for any type of modal presentation style, contrary to the following docs:

> When you want to present a view controller using a custom modal presentation type, set its modalPresentationStyle property to custom and assign an object that conforms to this protocol to its transitioningDelegate property. When you present that view controller, UIKit queries your transitioning delegate for the objects to use when animating the view controller into position.

https://developer.apple.com/documentation/uikit/uiviewcontrollertransitioningdelegate

Though the documentation says that you should set the presentation style to custom, the transitioning delegate is still queried (confusing, I agree).

Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: `UIModalPresentationCustom` only affects whether or not UIKit asks the transitioning delegate for a custom presentation controller.

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.h
Comment (ianegordon) [reply components/MaskedTransition/src/private/MDCMaskedPresentationController.h]: If the presentationMode is not custom, I would not expect custom presentation elements.
My understanding is the UIPresentationController (iOS 8) was created to overcome shortcoming in the transitioningDelegate system (iOS 7).
I don't think that the added complexity of exposing this scrim through CreateScrimView is a good idea, especially as the scrimView is owned by either the context or the _presentationLayer.  I would expect the owner to control the add/remove and alpha of any child views.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: originalFrame may not be standardized.
Use CGRectGetWidth()/Height()/MinY()/MinX.
If you decide to CGRectStandardize, please rename originalStandardizedFrame so this will not raise red flags.

**Review by ianegordon — CHANGES_REQUESTED** [review]

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.h
Comment (jverkoey) [reply components/MaskedTransition/src/private/MDCMaskedPresentationController.h]: The background dimming behavior is part of the motion spec for full-screen masked transitions. If we choose to deviate from the spec then I'm happy to remove the transition-owned scrim logic and the corresponding scrim APIs.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: Can you elaborate on the specific concern with standardization?

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.h
Comment (ianegordon) [reply components/MaskedTransition/src/private/MDCMaskedPresentationController.h]: Deviating from spec seems,...... unlikely? Extremely unlikely? 
Lets touch base in person and discuss the options.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: We are using multiple frames to create this rect.  
Some might have negative width / heights with an origin based of that, while others may have the standard positive width / heights.
Apple fixed this in Swift, but we still need to go through the CGRectFunc()s in Objective-c.

Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: I'm not certain I'm following what the specific problem here is - if `originalFrame`'s width/height are originally negative, I'm not certain that it's our transition's responsibility to correct that.

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.h
Comment (jverkoey) [reply components/MaskedTransition/src/private/MDCMaskedPresentationController.h]: To clarify: I'm willing to push for updating the spec to remove the scrim behavior on full-screen transitions. The purpose of a scrim is to indicate that underlying content can't be directly interacted with. For a full-screen transition, this purpose is nulled, so I can argue that the scrim's appearance is unhelpful.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: originalFrame having a negative .size.width isn't an error to be corrected, it's an acceptable state.  The following two rects are considered equal (CGRectEqualToRect()):
rectStandardized .origin = 10, 10, .size = 40, 20
rectNegativeSize .origin = 50, 30, .size = -40, -20
We should have the same behavior on standardized and non-standardized rects.

**Comment by jverkoey** [comment]
- [x] Explore using custom modal presentation with full screen transition and managing background view removal and viewDidDisappear case ourself.
- [x] Make sure to use CGRectGetWidth/Height in all places.

**Comment by jverkoey** [comment]
Have made the requested changes. PTAL.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (samnm) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: If you assign the sourceView to the presentation controller when you instantiate it then you can avoid having to hold a reference to the presentation controller. I might be wrong but I think the expected pattern is to not hold a reference to your presentation controller.

**Comment by jverkoey** [comment]
Gentle ping.

**Review by ajsecord — APPROVED** [review]

**Comment by jverkoey** [comment]
@ianegordon less-gentle ping. Needs approval before I can land.

**Comment by jverkoey** [comment]
Some follow-up todos:

- [x] Bug: Fix rotation layout for presented dialog.
- [x] How do I use this presentation controller with a transitioning delegate?
- [x] Bug: Look into toolbar motion timing.
- [ ] Make some on-device recordings for motion design review.
- [ ] File an issue: Make it possible to slow down animations.
- [ ] Important feature: Support shadows on the button.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [inline components/MaskedTransition/src/MDCMaskedTransition.m]: Use hypot()

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (ianegordon) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: If you are unable to fix this now, please create an issue and add a TODO.

**Review by ianegordon — CHANGES_REQUESTED** [review]
Please address earlier comment, how do I use this presentation controller with a transitioning delegate?
See BottomSheet and Dialog for our previous implementation.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: Done.

File: components/MaskedTransition/src/MDCMaskedTransition.m
Comment (jverkoey) [reply components/MaskedTransition/src/MDCMaskedTransition.m]: Done.

**Comment by jverkoey** [comment]
> Please address earlier comment, how do I use this presentation controller with a transitioning delegate? See BottomSheet and Dialog for our previous implementation.

Will look into this.

**Comment by jverkoey** [comment]
> RE: How do I use this presentation controller with a transitioning delegate?

The Transitioning library provides a small abstraction over the transitioning delegate API in the form of the `mdm_transitionController` API available on all UIViewController instances. This transition controller acts as a glue between `Transition` objects and UIKit's transitioning APIs.

If someone did not want to use this glue but still wanted to use the `Transition` instance, they would do the following:

1. Implement the transitioning delegate.
2. Implement the `TransitionContext` protocol, providing implementations for all of the required APIs.
3. Create an instance of the desired `Transition` type and invoke `startWithContext:` on the transition.

The above is, in essence, what the transition controller instance is doing, though the transition controller is taking care to handle edge cases and initialization that would otherwise need to be re-implemented for each transition type.

**Comment by jverkoey** [comment]
Fixed the noted bugs and verified that the toolbar timing is as spec'd.

**Comment by jverkoey** [comment]
@ianegordon for final review

File: components/MaskedTransition/src/private/MDCMaskedPresentationController.m
Comment (ianegordon) [inline components/MaskedTransition/src/private/MDCMaskedPresentationController.m]: I'm seeing a CGFloat conversion error here.

**Review by ianegordon — CHANGES_REQUESTED** [review]

**Comment by ianegordon** [comment]
Tested locally on iPhone 5 running iOS 8.3.

**Comment by ianegordon** [comment]
Rendering Strangeness on iPhone 4S running 9.3.5.
Left Card Transition from the right button appear to jump off the left side of the screen.

- [x] Investigate

**Comment by jverkoey** [comment]
Fixed the iPhone 4S bug.

**Comment by ianegordon** [comment]
Verified on iPhone 4s @ 9.3.5.
Whoho!

**Review by ianegordon — APPROVED** [review]

**Pull request merged**
